sled-agent/src/dump_setup.rs

```diff
@@ -1661,4 +1702,436 @@ mod tests {
+            self
+        }
+    }
+
+    #[tokio::test]
+    async fn test_debug_dir_not_cleaned_under_threshold() {
+        let logctx = omicron_test_utils::dev::test_setup_log(
+            "test_debug_dir_not_cleaned_under_threshold",
+        );
+
+        const USED: u64 = 1024;
+        const AVAILABLE: u64 = 1024;
+
+        let mut files = TestDebugFileHarness::new(&logctx.log).await;
+
+        files.add_file("test1.log").set_size(512);
+        files.add_file("test2.log").set_size(512);
+        files.create_all();
+
+        // This is an assertion that "the test expectations are valid", before
+        // we actually try to do any cleanup.
+        //
+        // In this test in particular, it's also a bit of a no-op -- we aren't
+        // planning on deleting any files, because we're already below the
+        // threshold -- but it'll flag loudly if that threshold changes.
+        const CAPACITY: u64 = USED + AVAILABLE;
+        files.confirm_that_deleting_marked_files_will_drop_below_capacity(
+            USED, CAPACITY,
+        );
+
+        let worker = files.new_dump_setup_worker(USED, AVAILABLE).await;
+
+        // Before we cleanup: All files in "debug" exist
+        files.check_all_files_exist();
+        worker.cleanup().await.unwrap();
+
+        // After cleanup: All files still exist
+        files.check_all_files_exist();
+
+        logctx.cleanup_successful();
+    }
+
+    #[tokio::test]
+    async fn test_debug_dir_cleaned_over_threshold() {
+        let logctx = omicron_test_utils::dev::test_setup_log(
+            "test_debug_dir_cleaned_over_threshold",
+        );
+
+        const USED: u64 = 1024;
+        const AVAILABLE: u64 = 0;
+
+        // Create some test files.
+        //
+        // Their sizes sum to "USED".
+
+        let mut files = TestDebugFileHarness::new(&logctx.log).await;
+        files.add_file("test-1.log").set_size(256);
+        files.add_file("test-2.log").set_size(256);
+        files
+            .add_file("test-3.log")
+            .set_size(512)
+            .make_older()
+            .should_be_deleted();
+        files.create_all();
+
+        // This is an assertion that "the test expectations are valid", before
+        // we actually try to do any cleanup.
+        const CAPACITY: u64 = USED + AVAILABLE;
+        files.confirm_that_deleting_marked_files_will_drop_below_capacity(
+            USED, CAPACITY,
+        );
+
+        let worker = files.new_dump_setup_worker(USED, AVAILABLE).await;
+
+        // Before we cleanup: All files in "debug" exist
+        files.check_all_files_exist();
+        worker.cleanup().await.unwrap();
+
+        // After we cleanup: The files we marked as "should_be_deleted" are
+        // removed, but the rest still exist.
+        files.check_expected_files_removed();
+
+        logctx.cleanup_successful();
+    }
+
+    #[tokio::test]
+    async fn test_debug_dir_cleaned_over_threshold_many_files() {
```

Comment:
This tests that "the debug cleaner works correctly, even when there are files split across directories".